pynicotine/gtkgui/userbrowse.py

```diff
@@ -688,34 +712,277 @@ def on_select_dir(self, selection):
+                    # Single or Multi-selection
+                    self.on_upload_files()
+
+                elif int(len(self.file_store)) <= 0:
+                    # Avoid navigation trap
+                    self.on_upload_directory_recursive_to()
+                    self.FolderTreeView.grab_focus()
+
+            else:
+                if int(len(self.file_store)) >= 1 and (self.num_selected_files < 1):
+                    # No selection
+                    #self.on_download_directory()  # disabled to prevent accidental folder downloading
+                    return True
+
+                elif int(len(self.file_store)) >= 1 and (self.num_selected_files >= 1):
+                    # Single or Multi-selection
+                    self.on_download_files()
+
+                elif int(len(self.file_store)) >= 1 and (self.num_selected_files < 1):
+                    self.on_download_directory()
+
+                elif int(len(self.file_store)) <= 0:
+                    # Avoid navigation trap
+                    self.FolderTreeView.grab_focus()
+
+            return True
+
+        # Control the download_to location / Upload To
+        keycodes, mods = parse_accelerator("<Primary>Return")
+
+        if state & mods and keycode in keycodes:
+            self.select_files()
+
+            if self.num_selected_files >= 1 and int(len(self.file_store)) > 0:
+
+                if self.user == config.sections["server"]["login"]:
+                    self.on_upload_files()
+
+                else:
+                    self.on_download_files_to()
+
+                return True
+
+            elif self.num_selected_files < 1 and int(len(self.file_store)) > 0:
+
+                if self.user == config.sections["server"]["login"]:
+                    self.on_upload_directory_to()
+
+                else:
+                    self.on_download_directory_to()
+
+                return True
+
+            elif int(len(self.file_store)) <= 0:
+
+                self.FolderTreeView.grab_focus()
+
+                if self.user == config.sections["server"]["login"]:
+                    self.on_upload_directory_recursive_to()
+
+                else:
+                    self.on_download_directory_to()
+
+                return True
+
+        # Upload files to User shortcut
+        keycodes, mods = parse_accelerator("<Primary>u")
+
+        if state & mods and keycode in keycodes:
+            if self.user == config.sections["server"]["login"]:
+                self.select_files()
+                self.on_upload_files()
+                return True
+
+        # File properties
         keycodes, mods = parse_accelerator("<Alt>Return")
 
         if state & mods and keycode in keycodes:
             self.on_file_properties()
             return True
 
+        self.on_browse_key_press_event(*args)
+        return False
+
+    def on_browse_key_press_event(self, *args):
```

Comment:
This has been added here with the intention of reducing duplication, perhaps some of the above conditions can be re-worked to something more palatable. As these keys need a wider scope, perhaps there would be a better place for this?

Reply:
Would self.Main be appropriate?

Reply:
> Would self.Main be appropriate?

I don't think so no, because all the shortcuts are in reference to the particular user tab page.

> Add another key controller for a container widget

There is no container that has an id assignment, I tried giving a GtkBox an id, but this didn't work. Whilst it would be preferable to have shortcuts working whilst using Find, it's not a severe keyboard navigation trap as it's possible to escape out of the SearchEntry box with the down arrow key into the FolderTreeView. So I'm happy to leave this for the time being.

Reply:
Each tab has a separate self.Main container widget (see userbrowse.ui).

Reply:
I have tried adding a key_controller for selfMain into init, unfortunately it causes the Upload To User dialog escape function to be broken (other than that it works well).

`        self.key_controller = connect_key_press_event(self.Main, self.on_key_press_event_browse)`

Reply:
The solution that https://github.com/nicotine-plus/nicotine-plus/commit/a16bd4dc25d809241ac9ce3b63ab2abd518b6e1e offers makes this irrelevant now.